DOCSP-45179 Configure TLS (#129)
mongodb/docs-ruby: source/connect/tls.txt

.. _ruby-tls:

========================================
Configure Transport Layer Security (TLS) 
========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: security, authentication, transport layer security, encrypt

Overview
--------

In this guide, you can learn how to use the Transport Layer Security (TLS)
protocol to secure your connection to a MongoDB deployment.

To connect to a MongoDB deployment using TLS, you must:

- Enable a TLS connection in ``Mongo::Client``.
- Specify the client TLS certificate.
- Specify the CA certificate to verify the server's TLS certificate.

To learn how to configure your MongoDB deployment for TLS, see the
:manual:`TLS configuration guide </tutorial/configure-ssl/>` in the
{+mdb-server+} manual.

.. note::

   This page assumes prior knowledge of TLS/SSL and access to valid certificates.
   A full description of TLS/SSL, PKI (Public Key Infrastructure) certificates, and
   Certificate Authorities (CAs) is beyond the scope of this documentation. To

>    learn more about TLS, see the Wikipedia entry for :wikipedia:`Transport Layer Security <Transport_Layer_Security>`.

Author: I usually prefer to leave the period out of the hyperlink. Is there any style guidance on this?

.. _ruby-enable-tls:

Enable TLS
----------

You can enable TLS for the connection to your MongoDB deployment in the following ways:

- Set the ``ssl`` option to ``true`` in a new ``Mongo:Client`` object
- Set the ``tls`` option to ``true`` in your connection string

.. note:: SSL Naming Convention

    All {+mdb-server+} versions supported by the Ruby driver (2.6 and higher) 
    only implement TLS. 2.6 and do not use SSL.
    

>     For historical reasons, the Ruby option names pertaining to TLS configuration 
>     use the ``ssl`` rather than the ``tls`` prefix. The next major version of the Ruby 

Reviewer: s:

```suggestion
    For historical reasons, the Ruby driver prefixes TLS options with ``ssl`` instead of ``tls``.
    The next major version of the Ruby 
```

    driver (3.0) will use the ``tls`` prefix for Ruby option names.

.. tabs::

    .. tab:: Mongo::Client
        :tabid: mongoclient

        To configure certificates, you must specify the following options:

        - ``ssl_cert``: The certificate file used to identify the connection against MongoDB.
        - ``ssl_key``: The private keyfile used to identify the connection against MongoDB.
        - ``ssl_ca_cert``: The file containing the concatenated CA certificates 
          used to validate certs passed from the other end of the connection.

>           By default, the driver will use the default system root certificate 

Reviewer: i: [present tense](https://www.mongodb.com/docs/meta/style-guide/writing/use-present-tense/)

```suggestion
          By default, the driver uses the default system root certificate 
```

>           By default, the driver will use the default system root certificate 
>           store as the trust anchor, if the option is not designated. 

Reviewer: s:

```suggestion
          If you don't specify a value for this option, the driver will use the default system
          root certificate store as the trust anchor.
```

>         These certificate options can be specified as different types. For example, the 

Reviewer: s: instead of spelling this out, you could put the options in a table that includes a column for 'data type'

        ``ssl_cert`` option has the alternative ``ssl_cert_object`` and ``ssl_cert_string`` 
        options, to specify the certificate file as either a certificate object or a string.
        For more information on certificate options, see the Options Hash section of 
        the API Documentation for `Mongo::Client <{+api-root+}/Mongo/Client.html>`__ .  

        In the following example, the TLS certificate and corresponding private key 
        are provided in separate files:

        .. literalinclude:: /includes/connect/tls.rb
            :language: ruby
            :dedent:
            :start-after: start-enable-tls-settings
            :end-before: end-enable-tls-settings
        

>         The TLS certificate and private key can be contained in the same file, but 
>         both the certificate and private key options must still be specified. In the 

Reviewer: s: shortened

```suggestion
        You can specify both the TLS certificate and private key in a single file.
        In the 
```

Author: Thanks for the edit! Will keep in the part about still needing to specify both options, as I think it's important to clarify.

        following example, the TLS certificate and the private key are both defined 
        in the same ``client.pem`` file:

        .. literalinclude:: /includes/connect/tls.rb
            :language: ruby
            :dedent:
            :start-after: start-enable-tls-settings-same-file
            :end-before: end-enable-tls-settings-same-file

    .. tab:: Connection String
        :tabid: uri

        To configure certificates, you must specify the following URI options:
        
        - ``tlsCertificateKeyFile``: The certificate and key file used to identify

>           identify the connection against MongoDB. Both the certificate and key

Reviewer: ```suggestion
         the connection against MongoDB. Both the certificate and key
```

>           identify the connection against MongoDB. Both the certificate and key

Reviewer: q: is 'identify the connection against mongodb' the right language? i would have gone for 'identify the client to MongoDB' or something similar.

Author: 'identify the connection against mongodb' is the language used in the API documentation, but it is a bit strange sounding. Changed to 'verify the connection to MongoDB' .

>         - ``tlsCertificateKeyFile``: The certificate and key file used to identify
>           identify the connection against MongoDB. Both the certificate and key
>           must be stored in the same file provided to this option.

Reviewer: s:

```suggestion
        - ``tlsCertificateKeyFile``: The file that contains the certificate and key file
           used to identify the connection against MongoDB.
```

>         - ``tlsCAFile``: The file containing the concatenated certificate authority (CA)

Reviewer: ```suggestion
        - ``tlsCAFile``: The file containing the concatenated CA
```

          certificates used to validate certificates passed from the other end of the 

>           connection. The driver will use the default system root certificate 

Reviewer: s: does this mean server to client, client to server, or both? i would be specific.

Author: Changed to say 'from the Mongo deployment to the client', which I think is the correct flow. I can ask the technical reviewer to make 100% sure.

>           store as the trust anchor, if the option is not designated. 

Reviewer: s: could apply same suggestion as above

        .. literalinclude:: /includes/connect/tls.rb
            :language: ruby
            :dedent:
            :start-after: start-enable-tls-uri
            :end-before: end-enable-tls-uri
        
        The file containing the certificate and key usually has a``.crt`` or ``.pem`` 
        extension.

>         URI option values must be properly URI escaped. This applies, for example, 

Reviewer: nit: other docs call this 'percent-encoding' instead of 'URI escaped'

```suggestion
        URI option values must be percent-encoded. This applies, for example, 
```

        to slashes (/) in the paths, which are encoded as ``%2f``.

> Modify the SSL Context

Reviewer: q: this section talks only about TLS options. is 'Modify the TLS Context' a more appropriate heading?

Author: Since the method has TLS in it, sounds good!

----------------------

If your TLS configuration requires customization, you can set TLS context hooks

> by adding a native Ruby class ``Proc`` object to the ``Mongo.tls_context_hooks`` 

Reviewer: q: is 'Ruby class Proc object' the right description? i would have guessed 'native Ruby Proc object' or similar

Author: I think native Ruby Proc object' works, thanks!

> array. The ``Proc`` should be added to the array before creating any ``Mongo::Client`` 
> instances. 

Reviewer: i: use active voice

```suggestion
array. Add the ``Proc`` object to the array before you create any ``Mongo::Client`` 
instances. 
```

> For example, the following code could be placed in an initializer to enable the 
> AES256-SHA cipher as the only cipher to be used for TLS:

Reviewer: s:

```suggestion
The following code example enables the 
AES256-SHA cipher as the only cipher to be used for TLS:
```

.. literalinclude:: /includes/connect/tls.rb
    :language: ruby
    :dedent:
    :start-after: start-modify-context
    :end-before: end-modify-context

To learn more about the SSL context options available, see the Ruby
documentation for 

> `SSLContext <https://ruby-doc.org/stdlib-2.7.0/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html>`__.

Reviewer: nit:

```suggestion
`SSLContext. <https://ruby-doc.org/stdlib-2.7.0/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html>`__
```

Author: Same as above, I like to put periods outside the hyperlink since it's not a part of the context of the actual link! Please let me know if there is a documented standard, though

Reviewer: there's nothing in the style guide--just what i've seen as recommended. either way works though

OCSP Verification
-----------------

> If the certificate provided by the server contains an OCSP endpoint URI,

Reviewer: i: [first use of abbreviation](https://www.mongodb.com/docs/meta/style-guide/style/abbreviations/)

```suggestion
If the certificate provided by the server contains an Online Certificate Status Protocol (OCSP)
endpoint URI,
```

> the driver will issue an OCSP request to the specified endpoint to verify the

Reviewer: ```suggestion
the driver issues an OCSP request to the specified endpoint to verify the
```

> validity of the certificate by default.

Reviewer: s: move 'by default' to between the two clauses in this sentence. otherwise it seems to modify 'validity' or something else

> The OCSP endpoint check may be disabled by setting the
> ``:ssl_verify_ocsp_endpoint`` Ruby option to ``false`` or by setting the
> ``tlsDisableOCSPEndpointCheck`` URI option to ``true`` when creating a client.

Reviewer: i: passive voice

```suggestion
To disable the OCSP endpoint check, set the
``:ssl_verify_ocsp_endpoint`` Ruby option to ``false`` or set the
``tlsDisableOCSPEndpointCheck`` URI option to ``true`` when creating a client.
```

> .. note::

Reviewer: s: include note heading

```suggestion
.. note:: JRuby Stops OSCP Checks
```

Author: Thanks, changed to 'JRuby Does Not Support OCSP Verification'

>   OCSP endpoint checking is not currently performed when running on JRuby,
>   since JRuby does not correctly expose the OCSP endpoint URI.

Reviewer: s: indent note and reword for active voice

```suggestion
   Because JRuby doesn't correctly expose OCSP endpoint URIs,
   the driver doesn't check OCSP endpoints when the underlying
   application runs on JRuby.
  
```

        .. literalinclude:: /includes/connect/tls.rb
            :language: ruby
            :dedent:
            :start-after: start-enable-tls-uri
            :end-before: end-enable-tls-uri
        
        The file containing the certificate and key usually has a``.crt`` or ``.pem`` 
        extension.

        URI option values must be properly URI escaped. This applies, for example, 
        to slashes (/) in the paths, which are encoded as ``%2f``.

Modify the SSL Context
----------------------

If your TLS configuration requires customization, you can set TLS context hooks
by adding a native Ruby class ``Proc`` object to the ``Mongo.tls_context_hooks`` 
array. The ``Proc`` should be added to the array before creating any ``Mongo::Client`` 
instances. 

For example, the following code could be placed in an initializer to enable the 
AES256-SHA cipher as the only cipher to be used for TLS:

.. literalinclude:: /includes/connect/tls.rb
    :language: ruby
    :dedent:
    :start-after: start-modify-context
    :end-before: end-modify-context

To learn more about the SSL context options available, see the Ruby
documentation for 

> `SSLContext <https://ruby-doc.org/stdlib-2.7.0/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html>`__.

Author: For technical reviewer, is this an appropriate link to put for reference? I wasn't able to find much information on ``context`` options in our docs. Thanks!

   * - ``:ssl_ca_cert_object``
     - ``Array<OpenSSL::X509::Certificate>``
     - An array of objects representing the CA certificates used 
       to validate certificates passed from the MongoDB deployment to the client.
   * - ``:ssl_ca_cert_string``
     - ``String``
     - A string containing one PEM-encoded CA certificate used to validate certificates 
       passed from the MongoDB deployment to the client.

Modify the TLS Context
----------------------

If your TLS configuration requires customization, you can set TLS context hooks
by adding a native Ruby ``Proc`` object to the ``Mongo.tls_context_hooks`` 
array. Add the ``Proc`` object to the array before you create any ``Mongo::Client`` 
instances. 

The following code example enables the AES256-SHA cipher as the only cipher 
to be used for TLS:

.. literalinclude:: /includes/connect/tls.rb
    :language: ruby
    :dedent:
    :start-after: start-modify-context
    :end-before: end-modify-context

The {+driver-short+} TLS context options are based on native Ruby handling of SSL. To 
learn more about the TLS context options available, see the Ruby
documentation for 

> `SSLContext <https://ruby-doc.org/stdlib-2.7.0/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html>`__.

Author: For technical reviewer, is this an appropriate link to put for reference? I wasn't able to find much information on ``context`` options in our docs. Thanks!

Reviewer 2: Ruby 2.7 is EOL, so that link is probably not ideal. Linking out to the Ruby docs is otherwise probably fine, we just need to make sure the links remain relevant.

Here's the link to the same class, for Ruby 3.2.6: https://ruby-doc.org/3.2.6/exts/openssl/OpenSSL/SSL/SSLContext.html

   This page assumes prior knowledge of TLS/SSL and access to valid certificates.
   A full description of TLS/SSL, PKI (Public Key Infrastructure) certificates, and
   CAs is beyond the scope of this documentation. To learn more about TLS, see the 

>    Wikipedia entry for :wikipedia:`Transport Layer Security <Transport_Layer_Security>` .

Reviewer: ```suggestion
   Wikipedia entry for :wikipedia:`Transport Layer Security <Transport_Layer_Security>`.
```

        The file containing the certificate and key usually has a``.crt`` or ``.pem`` 
        extension.

        URI option values must be percent-encoded. This applies, for example, 
        to slashes (/) in the paths, which are encoded as ``%2f``.

Specify Client TLS Certificates
-------------------------------

The {+driver-short+} provides multiple options for you to specify
the TLS certificate, key, and CA certificate with different data or object types.

TLS Certificate
~~~~~~~~~~~~~~~

> You may provide one of the following options to specify the TLS certificate:

Reviewer: i: [may](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#m:~:text=Use%20may%20to%20indicate%20permission%20only.)

```suggestion
You can provide one of the following options to specify the TLS certificate:
```

.. list-table::
   :header-rows: 1
   :widths: 20 30 50

   * - Option Name
     - Data/Object Type Accepted
     - Description
   * - ``:ssl_cert``
     - ``String``
     - The certificate file path used to verify the connection to a MongoDB 
       deployment.
   * - ``:ssl_cert_object``
     - ``OpenSSL::X509::Certificate``

>      - The certificate object used to to verify the connection to a MongoDB 

Reviewer: ```suggestion
     - The certificate object used to verify the connection to a MongoDB 
```

       deployment. 
   * - ``:ssl_cert_string``
     - ``String``
     - A string containing the PEM-encoded certificate used to verify the connection to 
       a MongoDB deployment.

TLS Private Key
~~~~~~~~~~~~~~~

> You may provide one of the following options to specify the TLS private key:

Reviewer: ```suggestion
You can provide one of the following options to specify the TLS private key:
```

.. list-table::
   :header-rows: 1
   :widths: 20 30 50

   * - Option Name
     - Data/Object Type Accepted
     - Description
   * - ``:ssl_key``
     - ``String``
     - The private keyfile path used to verify the connection to a MongoDB 
       deployment.
   * - ``:ssl_key_object``
     - ``OpenSSL::PKey``
     - The private key object used to verify the connection to a MongoDB 
       deployment.
   * - ``:ssl_key_pass_phrase``
     - ``String``
     - A passphrase for the private key.
   * - ``:ssl_key_string``
     - ``String``
     - A string containing the PEM-encoded private key used to verify the 
       connection to a MongoDB deployment.

TLS CA Certificate
~~~~~~~~~~~~~~~~~~

> You may provide one of the following options to specify the TLS CA certificate:

Reviewer: ```suggestion
You can provide one of the following options to specify the TLS CA certificate:
```
